### PR TITLE
LPS-195615 - Allow admins translate filter names

### PR DESCRIPTION
**Pull request by juanjofgliferay**
Story https://liferay.atlassian.net/browse/LPS-186878
Task https://liferay.atlassian.net/browse/LPS-195615

## Problem
Filter name can be edited but it does not allow translations. 

## Solution
Needs a fresh DB
Under FF `LPS-172017` we create System objects that allow translations and use the `InputLocalized` component to deal with them. 
As a fallback, in case there is no match for the selected language, the fragment will return the field name used for the filter

### With FF enabled

[filter-label-w-ff-enabled2.webm](https://github.com/liferay-frontend/liferay-portal/assets/132653342/954cdef5-e3d3-4cd7-8aa6-8275b4b3a236)

### Without FF enabled

[filter-name-no-ff.webm](https://github.com/liferay-frontend/liferay-portal/assets/132653342/5157450d-62c4-4322-8166-86ca7dc04016)

### Dataset using fallback

[filter-name-w-wo-translation.webm](https://github.com/liferay-frontend/liferay-portal/assets/132653342/ba1704b2-fa63-4558-8ccc-aec26a753a3d)

Fix issue with the Fields tab when ff `LPS-172017` is not enabled

### CODE REVIEW

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Filters.tsx [hunk 1]
```diff
@@ -343,30 +373,50 @@ function AddFDSFilterModalContent({
...
+							id={nameFormElementId}
+							label={Liferay.Language.get('name')}
+							name="name"
+							onChange={(newFieldLabel) => {
+								setI18nFilterNames({
+									...i18nFilterNames,
+									...newFieldLabel,
+								});
+							}}
+							required
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Filters.tsx:389]: See how I implemented frontend validation in https://github.com/liferay-frontend/liferay-portal/pull/3659. It's based on `onBlur` events, enabling save button when entire form passes validation. If it has flaws, we can tweak it. We can even discuss a different behavior, that might have better UX. But in any case, we should implement the same solution in both places.

**Review by markocikos — COMMENTED** [review]
@juanjofgliferay please see comments inline

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Filters.tsx [hunk 2]
```diff
@@ -99,9 +104,16 @@ function AddFDSFilterModalContent({
 	const [fieldInUseValidationError, setFieldInUseValidationError] = useState<
 		boolean
 	>();
+	const [errorMessage, setErrorMessage] = useState('');
+	const fdsFilterNameTranslations = filter?.name_i18n || {
+		[defaultLanguageId]: filter?.name || '',
```
[the remark this answers is not in the document]
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Filters.tsx]: Fixed, not needed

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/portlet/FDSViewsPortlet.java [hunk 3]
```diff
@@ -589,28 +589,190 @@ private synchronized void _generate(
...
-		_createFDSClientExtensionFilterObjectDefintion(
-			fdsViewObjectDefinition, locale, userId);
-		_createFDSDateFilterObjectDefinition(
-			fdsViewObjectDefinition, locale, userId);
-		_createFDSDynamicFilterObjectDefintion(
-			fdsViewObjectDefinition, locale, userId);
-		_createFDSFieldObjectDefinition(
-			fdsViewObjectDefinition, locale, userId);
-		_createFDSSortObjectDefinition(fdsViewObjectDefinition, locale, userId);
+		ObjectDefinition fdsDynamicFilterObjectDefinition =
```
Comment (dsanz) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/portlet/FDSViewsPortlet.java:631]: we'll need to move all this to the corresponding methods `_create*****ObjectDefintion()`

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/FDSViewFragmentRenderer.java [hunk 4]
```diff
@@ -454,7 +454,17 @@ private JSONArray _getFiltersJSONArray(
 					).put(
 						"id", properties.get("fieldName")
 					).put(
-						"label", properties.get("name")
+						"label",
+						() -> {
+							String label = String.valueOf(
```
Comment (dsanz) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/FDSViewFragmentRenderer.java:459]: this seems candidate for a separate function that returns the filter label. We're using it twice, but we'll end up using it 3 times thanks to the ClientExtension filter PR, yet to come

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Filters.tsx [hunk 5]
```diff
@@ -83,14 +83,12 @@ function AddFDSFilterModalContent({
...
-		[defaultLanguageId]: filter?.name || '',
-	};
+	const fdsFilterLabelTranslations = filter?.label_i18n ?? {};
 	const [from, setFrom] = useState<string>(
 		(filter as IDateFilter)?.from ?? format(new Date(), 'yyyy-MM-dd')
 	);
-	const [i18nFilterNames, setI18nFilterNames] = useState(
-		fdsFilterNameTranslations
+	const [i18nFilterLabels, setI18nFilterLabels] = useState(
+		fdsFilterLabelTranslations
```
Comment (dsanz) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Filters.tsx:94]: nice!

**Review by dsanz — DISMISSED** [review]
We're close @juanjofgliferay , I think an extra round of changes will make it ready :+1:

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/portlet/FDSViewsPortlet.java (hunk 3 above)
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/portlet/FDSViewsPortlet.java:631]: Ouch

**Review by markocikos — CHANGES_REQUESTED** [review]
@juanjofgliferay see comments inline

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Filters.tsx [hunk 6]
```diff
@@ -312,30 +341,60 @@ function AddFDSFilterModalContent({
...
+								validateForm();
+							}}
+							onChange={setI18nFilterLabels}
+							required
+							translations={i18nFilterLabels}
+						/>
+					</ClayForm.Group>
+				) : (
+					<ClayForm.Group>
+						<label htmlFor={nameFormElementId}>
```
[the remark this answers is not in the document]
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Filters.tsx:373]: image
Should we replace the tooltip by an asterisk to identify the field as required?

[the remark this answers is not in the document]
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Filters.tsx:373]: Hmmm should we even make this required? :thinking:

[the remark this answers is not in the document]
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Filters.tsx:377]: After a lengthy discussion in https://liferay.slack.com/archives/C3JBR21HA/p1695650849062839, the conclusion is that we don't want to make the label field required. That should simplify code in this PR.

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Filters.tsx [hunk 7]
```diff
@@ -141,9 +150,24 @@ function AddFDSFilterModalContent({
...
 		let body: any = {
 			fieldName: selectedField.name,
-			name: name || selectedField.label,
 		};
 
+		if (Liferay.FeatureFlags['LPS-172017']) {
+			let labels;
+			if (!translationExists({translations: i18nFilterLabels})) {
+				const defaultLanguageId = Liferay.ThemeDisplay.getDefaultLanguageId();
+				labels = {[defaultLanguageId]: selectedField.label};
```
Comment (juanjofgliferay) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Filters.tsx:159]: Save fieldName as the default language in the same way we do without translations. We need the value to show in the filters table.
Screenshot from 2023-09-26 08-49-40

Another option is waiting for the filter list and modify the response if no label is available

Comment (dsanz) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Filters.tsx:159]: This is preferrable from a UI perspective. Considering that user can change labels at will, I don't see any reason not to implement this

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Filters.tsx [hunk 8]
```diff
@@ -542,6 +520,8 @@ function Filters({fdsFilterClientExtensions, fdsView, namespace}: IProps) {
 				OBJECT_RELATIONSHIP.FDS_VIEW_FDS_DYNAMIC_FILTER
 			] as ISelectionFilter[];
 
+			// TODO: set a default filter name
+
```
Comment (dsanz) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Filters.tsx:524]: Does this mean we want to do the same here? (the same = use field name as the filter label and save it if no label is provided

Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Filters.tsx:524]: :grimacing: It is related to the other change and forgot to remove the comment

**Comment by juanjofgliferay** [comment]
ci:test:relevant

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/FDSViewFragmentRenderer.java [hunk 9]
```diff
@@ -574,6 +569,19 @@ private JSONArray _getItemsActionsJSONArray(
 			});
 	}
 
+	private String _getLabelFromProperties(
+		Map<String, Object> fdsFieldProperties, String defaultLabel,
+		String fallbackLabel) {
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/FDSViewFragmentRenderer.java:574]: nit: For clarity, I'd call these `defaultProperty` and `fallbackProperty`

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Fields.tsx [hunk 10]
```diff
@@ -467,15 +467,17 @@ const EditFDSFieldModalContent = ({
...
 
 	const validateFDSField = function () {
-		const defaultLanguageId = Liferay.ThemeDisplay.getDefaultLanguageId();
+		if (Liferay.FeatureFlags['LPS-172017']) {
+			const defaultLanguageId = Liferay.ThemeDisplay.getDefaultLanguageId();
 
-		if (!i18nFieldLabels[defaultLanguageId]) {
-			setErrorMessage(Liferay.Language.get('required'));
+			if (!i18nFieldLabels[defaultLanguageId]) {
+				setErrorMessage(Liferay.Language.get('required'));
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Fields.tsx:474]: Outside of scope of this task, but we should remove required from fields as well.

**Review by markocikos — COMMENTED** [review]
@juanjofgliferay please see comment inline

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/FDSViewFragmentRenderer.java (hunk 9 above)
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/FDSViewFragmentRenderer.java:574]: :+1: I will also change the order of the parameters to keep them alphabetical ordered

**Comment by juanjofgliferay** [comment]
@markocikos then, I'm fine leaving it empty to inform the admin about the issue.
I think that we cannot get rid of the `_getLabelFromProperties` because the default language used to create the label may not be the one used when displaying the dataset, so the fragment should take care of not showing a empty label filter

**Review by markocikos — APPROVED** [review]
@juanjofgliferay LGTM! :sparkles:

**Comment by juanjofgliferay** [comment]
ci:test:relevant

**Comment by dsanz** [comment]
@mateusralv , would you mind checking the relevant and stable, and let us know if they're related to this PR? Thanks a lot

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Filters.tsx [hunk 11]
```diff
@@ -304,38 +315,51 @@ function AddFDSFilterModalContent({
 							<DateRangeFilterModalContent.Header />
 						)}
 
-						{filterType !== EFilterType.SELECTION && (
-							<DateRangeFilterModalContent.Header />
+						{filterType === EFilterType.SELECTION && (
+							<SelectionFilterModalContent.Header />
```
Comment (dsanz) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Filters.tsx:319]: :+1:

**Comment by CarlosBrichete** [comment]
Hey @juanjofgliferay @dsanz, failures seems to not be related but after the last 3 commits the stable suite is failing, so I would run it again

**Comment by juanjofgliferay** [comment]
ci:test:stable

**Comment by juanjofgliferay** [comment]
ci:forward:force

**Comment by juanjofgliferay** [comment]
ci:forward:force

**Comment by juanjofgliferay** [comment]
Manual forward [to Brian https://github.com/brianchandotcom/liferay-portal/pull/141399](https://github.com/brianchandotcom/liferay-portal/pull/141399)

**Pull request closed without merging**